[JIT] Add selective backend lowering API

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43613 [JIT] Add selective backend lowering API**

**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

Commit:
[JIT] Add selective backend lowering API

**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

[ghstack-poisoned]

SplitInfinity commented on torch/jit/_backends.py:
```diff
@@ -0,0 +1,63 @@
+            lowered_submodule = to_backend_fn(submodule)
+            module_jit_type = module._c._type()
+            # Modify the corresponding attribute on the owning module's type.
+            module_jit_type.unsafe_remove_attribute(name)
+            module_jit_type.add_attribute(name, lowered_submodule._c._type())
```
One thing that worries me is that by default, types are shared across different calls to `torch.jit.script`. So if a user has multiple instances of `CoolModule`, scripts them all and selectively lowers a submodule of one of them, then this will modify the JIT type of all `CoolModule` instances. I see two options:

1) Let the user explicitly opt out of type sharing if they want to do this (probably a bad idea)
2) Try to mimic some sort of copy-on-write behaviour where we clone the type we want to edit just before we do this.

suo replied:
I think it would be safer/easier to just copy the entire module hierarchy and perform the edits on that. That's the approach that quantization takes, I believe there is already cloning + typemapping logic hanging off `Module` somewhere.

SplitInfinity commented:
Ping @suo
cc: @jackm321

SplitInfinity commented:
Ping @suo
cc: @jackm321

suo commented on torch/jit/_backends.py:
```diff
@@ -0,0 +1,63 @@
+    base: str,
+) -> RecursiveScriptModule:
+    # For each submodule:
+    for name, submodule in module._modules.items():
```
In general, we should implement logic like this is in C++, to avoid increasing the surface area of the Python bindings and re-implementing functionality like `remapTypes`.

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -2,11 +2,65 @@
+    // Adjust the parent's type so that the type of the submodule matches
+    // the type of lowered_submodule.
+    auto parent_jit_type = parent.type();
+    parent_jit_type->unsafeRemoveAttribute(qual_module_name.name());
```
This changes the ordering of the attributes in `parent_jit_type`, which might be surprising (since our slot representations are ordered).

SplitInfinity replied:
Ah that's true, so should I remove all of them and add them back? Or since we're doing unsafe type mods anyway, I'm going to add an `unsafeSwapAttribute` so we can override the `ClassAttribute` struct for that slot instead of adding and removing.

suo replied:
I think if you do the `type_remapper` thing I suggested below, you don't need to add anything new, since the remapping will happen while the type is cloned (and thus ordering will be preserved)

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -234,6 +288,30 @@ void initJitBackendBindings(PyObject* module) {
+          // interfere with the changes that to_backend_selective_impl will
+          // make.
+          Module& mod = original_module.value();
+          auto cloned_mod = mod.clone();
```
Hm, so this creates a module hierarchy with all submodule types cloned, but then you go in and swap the submodule types again to the lowered version. I think that will leave dangling references to the cloned submodule types in various method graphs. So if you had code that was like:
```
prim::CallMethod(%submod : ClonedType, "forward")
```
We haven't fixed up the cloned type to actually point to `LoweredType`. So the implementation of `CallMethod` will access the `ClonedType`'s forward in the interpreter instead of the new `LoweredType`'s version.

1. The type remapper in `clone_impl` is supposed to do the work of finding all references to type A and remap them to type B.
2. This suggests that we should write additional tests where the lowered and python versions of the module actually return *different* results, so that we can test that we are calling the right one.

SplitInfinity replied:
I added FileCheck tests for the post lowering graphs to make sure that types are replaced (although FileCheck is too coarse grained to do the kind of checks we need).

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

mortzur commented on test/jit/test_backends.py:
```diff
@@ -34,6 +36,13 @@ def to_test_backend_multi(module, method_compile_spec):
 
+def to_test_backend_selective(module, method_compile_spec, submodules):
+    def _to_test_backend(module):
+        return to_test_backend(module, method_compile_spec)
```
I haven't noticed that so far. This means that all submodules have have the same method_compile_spec. Right?
More accurately, this assumes that the logic to map a module to its spec is found in _to_test_backend method.
To do it nicely this would really be a dictionary - will a dictionary with scriptModule as key work? (a list in the same order of submodules can also be used, more error prone though)
Alternatively, "submodules" could be a dictionary between submodules strings and method_comile_spec dictionaries.
What do you think would be the best solution here?

SplitInfinity replied:
Dictionary from fully qualified submodule path to method_compile_spec could work. For example: `{'a.b.c' : spec1, 'd.e.f': spec2}` where `a.b.c` means attribute `c` of attribute `b` of attribute `a` of `module`.

This would be all up to you because you can define `_to_{backend}`, the Python function you pass to `torch._C. _jit_to_backend_selective` however you want as long as it accepts one `ScriptModule` as an argument. In this example, my implementation of `_to_test_backend` captures `method_compile_spec` from the enclosing scope and uses the same one for all lowered modules but it doesn't have to be that way.

SplitInfinity commented:
PIng @suo

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -2,11 +2,83 @@
+      return it->second;
+    };
+
+    for (auto& fn : parent_type->methods()) {
```
Still not totally sure about the type remapping logic here. For example, this assumes that only the direct parent of a module needs to have the graphs fixed up, but a module can appear in the graph of any of its recursive parents (through an attribute access like: `self.nested1.nested2.lowered_module`).

So I think to get all the corner cases, it's best to specify what should happen top-down and apply the transformation globally. e.g. build up the typemap of all modules to be lowered, the walk the module hierarchy and rewrite all methods/slots according to that map.

SplitInfinity replied:
I modified `to_backend_selective_impl` so that it maintains references to all of the ancestors of the lowered module, fixes up the parent type (the type of the direct ancestor), and then remaps types the graphs and schemas of all of the methods of all of its ancestors.

Do we need to worry about rewriting slots in any module except the parent? A module can appear only once in the hierarchy (for now, lol), and I handled the type remapping in the parent's slots using `unsafeChangeAttributeType` which doesn't change slot ordering.

Also, does the IR include specific slot numbers/indexes? I've only seen `getAttr` calls in IR that are [translated to slot numbers](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/interpreter.cpp#L801) at runtime [when `InterpreterState` is created](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/graph_executor.cpp#L390). So I don't think modifying the slot for an attribute after the IR has been generated will lead to problems. However, the return type of that `getAttr` still needs to be remapped after lowering.

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

SplitInfinity commented:
I added tests for errors that should be thrown when lowering something that's not a module and lowering a hierarchy that doesn't have unique types.

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

SplitInfinity commented:
Ping @suo

SplitInfinity commented:
Ping @suo

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -2,11 +2,119 @@
+// qualified names of submodules of \p mod that should be lowered.
+void toBackendSelectiveImpl(
+    Module& mod,
+    py::function to_backend,
```
lint seems legit

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -2,11 +2,119 @@
+
+    // Check that the parent type is not shared and therefore can be edited.
+    if (duplicate_types.count(parent.type()) > 0) {
+      throw py::cast_error(c10::str(
```
Possibly: can we add which module is actually shared, for debug purposes?

suo reviewed (approved):
Okay, lgtm! Thanks for the changes.

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add selective backend lowering API"


**Summary**
This commit adds a helper/utility to faciliate the selective lowering of
specific submodules within a module hierarchy to a JIT backend. The reason
that this is needed is that lowering a submodule of a scripted
module to a backend after the module has been scripted requires
adjusting its JIT type.

**Test Plan**
This commit refactors `NestedModuleTest` in `jit/test_backends.py` to
use this new selective lowering API.

**Fixes**
This commit fixes ##41432.

Differential Revision: [D23339855](https://our.internmc.facebook.com/intern/diff/D23339855)

[ghstack-poisoned]
